src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+            int l = node.transition(0);
+            int r = node.transition(childCount - 1);
+            assert 0 <= l && l <= r && r <= 255;
+            dest.writeByte(l);
+            dest.writeByte(r - l);      // r is included, i.e. this is len - 1
+
+            for (int i = 0; i < childCount; ++i)
+            {
+                int next = node.transition(i);
+                while (l < next)
+                {
+                    writeBytes(dest, NULL_VALUE);
+                    ++l;
+                }
+                writeBytes(dest, -node.serializedPositionDelta(i, nodePosition));
+                ++l;
+            }
+        }
+    };
+
+    static final Dense12 DENSE_12 = new Dense12();
+    static class Dense12 extends Dense
+    {
+        // byte flags
+        // byte start
+        // byte length-1
+        // length 12-bits transition targets (-1 for not present)
+        // var payload
+
+        Dense12()
+        {
+            super(FRACTIONAL_BYTES);
+        }
+
+        @Override
+        public int payloadPosition(ByteBuffer src, int position)
+        {
+            return position + 3 + (transitionRange(src, position) * 3 + 1) / 2;
+        }
+
+        @Override
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            return -read12Bits(src, position + 3, childIndex);
+        }
+
+        @Override
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            int l = node.transition(0);
+            int r = node.transition(node.childCount() - 1);
+            return 3 + ((r - l + 1) * 3 + 1) / 2;
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            int childCount = node.childCount();
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+            int l = node.transition(0);
+            int r = node.transition(childCount - 1);
+            assert 0 <= l && l <= r && r <= 255;
+            dest.writeByte(l);
+            dest.writeByte(r - l);      // r is included, i.e. this is len - 1
+
+            int carry = 0;
+            int start = l;
+            for (int i = 0; i < childCount; ++i)
+            {
+                int next = node.transition(i);
+                while (l < next)
+                {
+                    carry = write12Bits(dest, NULL_VALUE, l - start, carry);
+                    ++l;
+                }
+                long pd = node.serializedPositionDelta(i, nodePosition);
+                carry = write12Bits(dest, (int) -pd, l - start, carry);
+                ++l;
+            }
+            if (((l - start) & 1) == 1)
+                dest.writeByte(carry);
+        }
+
+        @Override
+        boolean fits(long delta)
+        {
+            return 0 <= delta && delta <= 0xFFF;
+        }
+    };
+
+    static final LongDense LONG_DENSE = new LongDense();
```

Comment:
nit: Should probably move this into `LongDense` itself to avoid potential classloading deadlocks